File: Python/utils/linked_list.py
from __future__ import annotations

import copy
from typing import Any


class ListNode:
    def __init__(self, val: Any, next:ListNode|None=None):
        self.val = val
        self.next = next

    @classmethod
    def from_list(cls, nodes: list[Any]) -> ListNode|None:
        if 0 == len(nodes):
            return None
        head: ListNode = ListNode(nodes[0])
        tail: ListNode = head
        for node in nodes[1:]:
            tail.next = ListNode(node)
            tail = tail.next
        return head

    def to_list(self) -> list[Any]:
        nodes: list[Any] = []
        head: ListNode = copy.deepcopy(self)
        while head is not None:
            nodes.append(head.val)
            head = head.next
        return nodes

    def __repr__(self):
        return str(self.to_list())

    def __eq__(self, other):
        if self is None and other is None:
            return True
        elif self is None or other is None:
            return False
        return self.to_list() == other.to_list()

File: TypeScript/utils/linked-list.ts
export class ListNode {
    val: any;
    next: ListNode | null;

    constructor(val: any, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }

    static fromList(nodes: any[]): ListNode | null {
        if (nodes.length === 0) return null;
        const head = new ListNode(nodes[0]);
        let tail = head;
        for (let i = 1; i < nodes.length; i++) {
            tail.next = new ListNode(nodes[i]);
            tail = tail.next;
        }
        return head;
    }

    toList(): any[] {
        const nodes: any[] = [];
        let head: ListNode | null = this;
        while (head !== null) {
            nodes.push(head.val);
            head = head.next;
        }
        return nodes;
    }

    toString(): string {
        return JSON.stringify(this.toList());
    }

    equals(other: ListNode | null): boolean {
        if (other === null) return false;
        return JSON.stringify(this.toList()) === JSON.stringify(other.toList());
    }
}
